### PR TITLE
Model adapter controller improvement and refactor

### PR DESCRIPTION
**Pull request by Jeffwan**
## Pull Request Description

1. Add artifact url part of the spec, we do not use untyped string from additonalConfig anymore.
2. Add more validation (code + kubebuilder tag) for ModelAdapter fields
3. Refactor the workflow and extract some utils for easy testing
4. Add more tests against those utilities

## Related Issues
Resolves: #41 #51 part of #136

**Important: Before submitting, please complete the description above and review the checklist below.**

---

<details>
<summary><strong>Contribution Guidelines (Expand for Details)</strong></summary>

<p>We appreciate your contribution to aibrix! To ensure a smooth review process and maintain high code quality, please adhere to the following guidelines:</p>

<h3>Pull Request Title Format</h3>
<p>Your PR title should start with one of these prefixes to indicate the nature of the change:</p>
<ul>
    <li><code>[Bug]</code>: Corrections to existing functionality</li>
    <li><code>[CI]</code>: Changes to build process or CI pipeline</li>
    <li><code>[Docs]</code>: Updates or additions to documentation</li>
    <li><code>[API]</code>: Modifications to aibrix's API or interface</li>
    <li><code>[CLI]</code>: Changes or additions to the Command Line Interface</li>
    <li><code>[Misc]</code>: For changes not covered above (use sparingly)</li>
</ul>
<p><em>Note: For changes spanning multiple categories, use multiple prefixes in order of importance.</em></p>

<h3>Submission Checklist</h3>
<ul>
    <li>[ ] PR title includes appropriate prefix(es)</li>
    <li>[ ] Changes are clearly explained in the PR description</li>
    <li>[ ] New and existing tests pass successfully</li>
    <li>[ ] Code adheres to project style and best practices</li>
    <li>[ ] Documentation updated to reflect changes (if applicable)</li>
    <li>[ ] Thorough testing completed, no regressions introduced</li>
</ul>

<p>By submitting this PR, you confirm that you've read these guidelines and your changes align with the project's contribution standards.</p>

</details>

### CODE REVIEW

File: pkg/controller/modeladapter/resources.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		{
+			Addresses: []string{pod.Status.PodIP},
+		},
+	}
+
+	ports := []discoveryv1.EndpointPort{
+		{
+			Name:     stringPtr("http"),
+			Protocol: protocolPtr(corev1.ProtocolTCP),
+			Port:     int32Ptr(8000),
```
Comment (varungup90) [inline pkg/controller/modeladapter/resources.go:42]: Port number, I am also using static value. Should we make it a const, then how to enforce it at deployment OR should we always read pod's port value from PodSpec?

File: pkg/controller/modeladapter/modeladapter_controller.go [hunk 2]
```diff
@@ -349,26 +341,28 @@ func (r *ModelAdapterReconciler) DoReconcile(ctx context.Context, req ctrl.Reque
...
+		if err = r.updateStatus(ctx, instance, instance.Status); err != nil {
+			return reconcile.Result{}, fmt.Errorf("update modelAdapter status error: %v", err)
 		}
 	}
 
 	return ctrl.Result{}, nil
 }
 
-func (r *ModelAdapterReconciler) clearModelAdapterInstanceList(ctx context.Context, instance *modelv1alpha1.ModelAdapter) (ctrl.Result, error) {
+func (r *ModelAdapterReconciler) updateStatus(ctx context.Context, instance *modelv1alpha1.ModelAdapter, status modelv1alpha1.ModelAdapterStatus) error {
```
Comment (varungup90) [inline pkg/controller/modeladapter/modeladapter_controller.go]: status is passed as an argument but not used. You can refactor the code to use helper updateStatus function at multiple places in this PR or next PR.

File: pkg/controller/modeladapter/modeladapter_controller.go [hunk 3]
```diff
@@ -349,26 +341,28 @@ func (r *ModelAdapterReconciler) DoReconcile(ctx context.Context, req ctrl.Reque
...
+	meta.SetStatusCondition(&instance.Status.Conditions, metav1.Condition{
+		Type:               string(modelv1alpha1.ModelAdapterConditionReady),
+		Status:             metav1.ConditionTrue,
+		Reason:             "Reconciling",
+		Message:            fmt.Sprintf("ModelAdapter %s is ready", klog.KObj(instance)),
+		LastTransitionTime: metav1.Now(),
+	})
+
+	return r.Status().Update(ctx, instance)
```
Comment (varungup90) [inline pkg/controller/modeladapter/modeladapter_controller.go:362]: You are updating the status here, you can also re-fetch here itself.

File: pkg/controller/modeladapter/resources.go (hunk 1 above)
Comment (Jeffwan) [reply pkg/controller/modeladapter/resources.go:42]: yeah, that's something we should consider in next phase, 8000 is fixed for vLLM. We need to support other frameworks and there's also no way to force user to use fixed port. We can use a const for short term. Let's create a separate issue to change all the references

File: pkg/controller/modeladapter/modeladapter_controller.go (hunk 2 above)
Comment (Jeffwan) [reply pkg/controller/modeladapter/modeladapter_controller.go]: let me make the change

File: pkg/controller/modeladapter/modeladapter_controller.go (hunk 3 above)
Comment (Jeffwan) [reply pkg/controller/modeladapter/modeladapter_controller.go:362]: sounds good. will do

File: pkg/controller/modeladapter/modeladapter_controller.go (hunk 2 above)
Comment (Jeffwan) [reply pkg/controller/modeladapter/modeladapter_controller.go]: Unused field has been corrected and I leave adopting helper to next PR.  the condition message needs to be carefully designed. If we use `updateStatus`, it needs to be changed.

File: pkg/controller/modeladapter/modeladapter_controller.go (hunk 3 above)
Comment (Jeffwan) [reply pkg/controller/modeladapter/modeladapter_controller.go:362]: Just notice this method is being used at the end, we do not need to re-fetch here. Once the util is adopted in some where else later, we can decide to re-fetch in the workflow.

File: pkg/controller/modeladapter/modeladapter_controller.go [hunk 4]
```diff
@@ -402,125 +397,137 @@ func (r *ModelAdapterReconciler) schedulePod(ctx context.Context, instance *mode
...
+	instance.Status.Phase = modelv1alpha1.ModelAdapterBinding
+	meta.SetStatusCondition(&instance.Status.Conditions, metav1.Condition{
+		Type:               string(modelv1alpha1.ModelAdapterConditionTypeScheduled),
+		Status:             metav1.ConditionTrue,
+		Reason:             "Reconciling",
+		Message:            fmt.Sprintf("ModelAdapter %s is loaded", klog.KObj(instance)),
+		LastTransitionTime: metav1.Now(),
+	})
+	if err := r.Status().Update(ctx, instance); err != nil {
+		klog.InfoS("Got error when updating status", "error", err, "ModelAdapter", instance)
```
Comment (varungup90) [inline pkg/controller/modeladapter/modeladapter_controller.go:439]: ErrorS? I noticed there are few more places where Info is used for logging error.

Comment (Jeffwan) [reply pkg/controller/modeladapter/modeladapter_controller.go:439]: nice catch. it should be error.  Let's have a clean up in this story https://github.com/aibrix/aibrix/issues/69

**Pull request merged**